### ISSUE TITLE
[Kotlin] Json validation flow

### ISSUE THREAD

**Issue by raderio** [square/moshi#814]
Usually on deserialization we want to collect all validation errors and return them in response to the client. 
In order to catch all error the validation should be done before object creation.

If it will be done after object creation, in `init` block, than validation will be separated in 2 phases, first one will be validation regarding non-nullable field, because we cannot create objects if they have non-nullable properties. 
The second one will be to do structural validation: min/max values, length, regexp pattern, etc.

Example
```
class Data(val name: String, val age: Int, val email: String)
```
when we send a json `{"email": "xyz@xyz.com", "name": "A"}` we should return a response where to indicate that `age` can not be null and `name` size must be greater than 2, at once to send all the errors.

```
class SignUpDto(val email: String, val password: String)
```
If we will send `{ "email": null, "password": null }` or `{}` to the endpoint, kotlin will throw an exception that ONE field is null, so we will be able in response only say that ONE field can not be null. What I want to achieve is that we should return in response that BOTH/ALL fields can not be null.
So, this can be done if validation will be done before object creation.

Also, structural validation will be great to do based on annotations and not in `init` block, something like JSR 303.
This will be useful for documentation generation, it is far easier to generate based on annotations.

How I see the flow:
1. Read annotations and generate validation rules. Can be done by 3rd party library. As result Map<Class, List<Rules>>
2. Parse json as a map
3. Apply validation rules. Validation rules can be passed as parameter in configuration.
4. Create object

This flow also can be used for validation in Java.

**Comment by swankjesse** [comment]
We're unlikely to do this in Moshi itself. But you can write a JsonAdapter.Factory that does validation after delegating to Moshi.

**Comment by JakeWharton** [comment]
Why did you close #780 and open a new issue with the same comment?

**Comment by raderio** [comment]
@swankjesse 
Is it possible to do validation before Moshi?

@JakeWharton 
In that issue the entire idea is split in messages, so I want to present an entire picture.

**Comment by swankjesse** [comment]
@raderio sure, I don't care what you do before calling Moshi.

I don't think there's anything here we want to change for Moshi. Doing validation in the data binding phase is always going to be unsatisfying:

 - good validation collects multiple errors. Moshi doesn't have enough context to know when to keep going vs. halt.

 - good validation needs context. Is this value allowed for this user? For their API version? For this environment?

 - good validation requires nontrivial business logic. The minimum amount of a payment might depend upon its currency, for example.

For these reasons I think annotation-based validation is bad. Further I don't think Moshi is the right place for validation beyond data binding: are the fields known and of the right types.

**Comment by raderio** [comment]
Annotation-based validation is good when you want to generate API documentation, for example for Swagger.
